### ISSUE TITLE
🐛 Bug Report: authentication is broken when scaling out the backend

### ISSUE THREAD

**Issue by famibo** [backstage/backstage#26903]
### 📜 Description

When scaling out the backend and moving Backstage backend plugins to a separate backend instance following error occurs: AuthenticationError - Missing credentials. We use Azure Graph for authentication.

### 👍 Expected behavior

When moving build-in Backstage backend plugins to a separate backend instance there shouldn't be an authentication error

### 👎 Actual Behavior with Screenshots

We have tried to move following Backstage backend plugin to a separate backend instance:
@backstage/plugin-kubernetes-backend/alpha

There is an authentication error 401 when frontend tries to call this moved plugin -> see reproduction steps

### 👟 Reproduction steps

Here is what we've done so far:

1. In the main config (in the main backend, running on port 7007)

```
discovery:
  endpoints:
    - target: http://localhost:7008/api/kubernetes
      plugins: [kubernetes]
```

3. in apis.ts (in the frontend app):

```
 createApiFactory({
    api: discoveryApiRef,
    deps: { configApi: configApiRef },
    factory: ({ configApi }) => FrontendHostDiscovery.fromConfig(configApi),
  }),
```

4. In the config of the second backend (running on port 7008):

```
backend:
  auth:
    externalAccess:
      - type: jwks
        options:
          url: http://localhost:7007/api/auth/.well-known/jwks.json
          issuer: http://localhost:7007/api/auth
```

5. We can see that the frontend calls now the kubernetes backend plugin in the second backend but we are getting error 401:
Request URL: http://localhost:7008/api/kubernetes/clusters

```
{
    "error": {
        "name": "AuthenticationError",
        "message": "Missing credentials",
        "stack": "AuthenticationError: Missing credentials\n    at DefaultHttpAuthService.credentials (/Users/artur/Projects/Backstage/devhub-app-backstage/node_modules/@backstage/backend-defaults/src/entrypoints/httpAuth/httpAuthServiceFactory.ts:150:13)\n
}
```

### 📃 Provide the context for the Bug.

Interestingly enough there is exactly the same problem when we move e.g. the @backstage/plugin-devtools-backend plugin to the second backend.
When frontend makes then a call to http://localhost:7008/api/devtools/info there is exactly the same 401 error as shown above. So this seems to be a general problem when mowing standard backstage backend plugins to a separate backend.

### 🖥️ Your Environment

```
% yarn backstage-cli info
yarn run v1.22.21
OS:   Darwin 23.6.0 - darwin/x64
node: v20.11.0
yarn: 1.22.21
cli:  0.27.1 (installed)
backstage:  1.31.1

Dependencies:
  @backstage/app-defaults                                          1.5.11
  @backstage/backend-app-api                                       0.5.14, 0.7.9, 0.9.2, 0.9.3, 1.0.0
  @backstage/backend-common                                        0.19.10, 0.20.2, 0.21.7, 0.23.3, 0.24.1, 0.25.0
  @backstage/backend-defaults                                      0.4.3, 0.4.4, 0.5.0
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.18
  @backstage/backend-plugin-api                                    0.6.21, 0.7.0, 0.8.0, 0.8.1, 1.0.0
  @backstage/backend-tasks                                         0.5.27
  @backstage/backend-test-utils                                    1.0.0
  @backstage/catalog-client                                        1.6.6, 1.7.0
  @backstage/catalog-model                                         1.6.0, 1.7.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.8, 0.2.7
  @backstage/cli                                                   0.27.1
  @backstage/config-loader                                         1.9.1, 1.9.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.15.0
  @backstage/core-compat-api                                       0.2.8, 0.3.0
  @backstage/core-components                                       0.14.10, 0.15.0
  @backstage/core-plugin-api                                       1.9.3, 1.9.4
  @backstage/dev-utils                                             1.1.0
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.9
  @backstage/frontend-plugin-api                                   0.6.7, 0.7.0, 0.8.0
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.30, 1.1.31
  @backstage/integration                                           1.14.0, 1.15.0
  @backstage/plugin-api-docs                                       0.11.9
  @backstage/plugin-app-backend                                    0.3.74
  @backstage/plugin-app-node                                       0.1.25
  @backstage/plugin-auth-backend-module-atlassian-provider         0.3.0
  @backstage/plugin-auth-backend-module-auth0-provider             0.1.0
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.2.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-server-provider  0.1.0
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.3.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.3.0
  @backstage/plugin-auth-backend-module-github-provider            0.2.0
  @backstage/plugin-auth-backend-module-gitlab-provider            0.2.0
  @backstage/plugin-auth-backend-module-google-provider            0.2.0
  @backstage/plugin-auth-backend-module-guest-provider             0.2.0
  @backstage/plugin-auth-backend-module-microsoft-provider         0.2.0
  @backstage/plugin-auth-backend-module-oauth2-provider            0.3.0
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.2.0
  @backstage/plugin-auth-backend-module-oidc-provider              0.3.0
  @backstage/plugin-auth-backend-module-okta-provider              0.1.0
  @backstage/plugin-auth-backend-module-onelogin-provider          0.2.0
  @backstage/plugin-auth-backend                                   0.23.0
  @backstage/plugin-auth-node                                      0.4.17, 0.5.0, 0.5.1, 0.5.2
  @backstage/plugin-auth-react                                     0.1.6
  @backstage/plugin-bitbucket-cloud-common                         0.2.22, 0.2.23
  @backstage/plugin-catalog-backend-module-azure                   0.2.2
  @backstage/plugin-catalog-backend-module-github                  0.7.3
  @backstage/plugin-catalog-backend-module-msgraph                 0.6.2
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.21, 0.2.0
  @backstage/plugin-catalog-backend                                1.26.0
  @backstage/plugin-catalog-common                                 1.0.26, 1.1.0
  @backstage/plugin-catalog-graph                                  0.4.9
  @backstage/plugin-catalog-import                                 0.12.3
  @backstage/plugin-catalog-node                                   1.12.5, 1.13.0
  @backstage/plugin-catalog-react                                  1.12.3, 1.13.0
  @backstage/plugin-catalog                                        1.23.0
  @backstage/plugin-devtools-backend                               0.4.0
  @backstage/plugin-devtools-common                                0.1.12
  @backstage/plugin-devtools                                       0.1.18
  @backstage/plugin-events-node                                    0.3.10, 0.3.9, 0.4.0
  @backstage/plugin-home-react                                     0.1.16, 0.1.17
  @backstage/plugin-home                                           0.7.10
  @backstage/plugin-kubernetes-backend                             0.18.6
  @backstage/plugin-kubernetes-common                              0.8.3
  @backstage/plugin-kubernetes-node                                0.1.19
  @backstage/plugin-kubernetes-react                               0.4.3
  @backstage/plugin-kubernetes                                     0.11.14
  @backstage/plugin-org                                            0.6.29
  @backstage/plugin-permission-backend                             0.5.49
  @backstage/plugin-permission-common                              0.7.14, 0.8.1
  @backstage/plugin-permission-node                                0.7.32, 0.8.1, 0.8.2, 0.8.3
  @backstage/plugin-permission-react                               0.4.25, 0.4.26
  @backstage/plugin-proxy-backend                                  0.5.6
  @backstage/plugin-scaffolder-backend-module-azure                0.1.15, 0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.13, 0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.13, 0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.13, 0.3.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.15, 0.2.0
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.13, 0.2.0
  @backstage/plugin-scaffolder-backend-module-github               0.4.1, 0.5.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.5, 0.5.0
  @backstage/plugin-scaffolder-backend                             1.24.0, 1.25.0
  @backstage/plugin-scaffolder-common                              1.5.5, 1.5.6
  @backstage/plugin-scaffolder-node-test-utils                     0.1.12
  @backstage/plugin-scaffolder-node                                0.2.10, 0.4.11, 0.4.9
  @backstage/plugin-scaffolder-react                               1.12.0
  @backstage/plugin-scaffolder                                     1.25.0
  @backstage/plugin-search-backend-module-catalog                  0.2.2
  @backstage/plugin-search-backend-module-pg                       0.5.35
  @backstage/plugin-search-backend-module-techdocs                 0.2.2
  @backstage/plugin-search-backend-node                            1.3.2
  @backstage/plugin-search-backend                                 1.5.17
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.8.0
  @backstage/plugin-search                                         1.4.16
  @backstage/plugin-signals-react                                  0.0.5
  @backstage/plugin-techdocs-backend                               1.10.13
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.14
  @backstage/plugin-techdocs-node                                  1.12.11
  @backstage/plugin-techdocs-react                                 1.2.8
  @backstage/plugin-techdocs                                       1.10.9
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.12
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.6.0
  @backstage/theme                                                 0.5.6, 0.5.7
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8, 1.0.9
✨  Done in 1.51s.
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

**Comment by benjdlambert** [comment]
Ahh this looks like it could be something to do with the `fetchApi` middleware that decorates the credentials to the upstream requests. I think by default it only will decorate requests that match the `backend.baseUrl`.

You have an option to override the `fetchApi` and provide your own implementation, but I wonder if there's something we should be contributing something to fix this.

Short term, I wonder if we could add a config option to allow for adding some URL's that we should decorate credentials with, but then long term I'm wondering if we should advise more usage of the `plugin://{pluginId}` schema, so that we can decorate all plugin requests automatically.

Just another thing, the `backend.auth.externalAccess` that you have configured in the second deployment shouldn't be necessary, you should remove that. You will probably want to setup a similar discovery config for the `auth` backend plugin to talk back to the other instance instead.

**Comment by freben** [comment]
One more note.

A critical part of the backend-to-backend auth system is that the receiver needs to be able to look up the address of who the caller claims to be, and check if it truly is the signer of the token. "Ok you claim to be plugin A wanting to talk to plugin B. Alright I am indeed B, but let's use the trusted DNS lookup mechanism to find an actual A node to see if their jwks really has a signing key for this".

So it's imperative that backend discovery is set up properly, if you segregate different backend plugins into different server instances. Check [this `backend.discovery.endpoints` config](https://github.com/backstage/backstage/blob/e2f0801e0cc5f8a78fc12026e3a0b39a59956759/packages/backend-defaults/config.d.ts#L559).

**Comment by freben** [comment]
Oh and one more thing. You seem to have very many duplicates of important packages - notably `@backstage/backend-app-api`. You may want to `yarn dedupe` your lockfile to not get version confusion issues.

The ideal `yarn backstage-cli info` output has exactly one version of every package, but you actually don't have to go quite that far. Mostly the `*app-api` packages are the most important to not duplicate. `yarn backstage-cli versions:bump` typically should help with this aspect when you do upgrades.

**Comment by famibo** [comment]
**@benjdlambert:** when I understand correctly your hint ("You will probably want to setup a similar discovery config for the auth backend plugin to talk back to the other instance instead") I've added following config to the second backend (running on port 7008):

discovery:
  endpoints:
    - target: http://localhost:7007/api/auth
      plugins: [auth]
      
 Also I've removed following config as suggested by you:
 
 backend:
  auth:
    externalAccess:
      - type: jwks
      
But unfortunately I'm still getting the same error 401 when frontend app calls the kubernetes backend plugin in the second backend:
Request URL: http://localhost:7008/api/kubernetes/clusters

{
    "error": {
        "name": "AuthenticationError",
        "message": "Missing credentials",
        "stack": "AuthenticationError: Missing credentials\n    at DefaultHttpAuthService.credentials (/Users/artur/Projects/Backstage/devhub-app-backstage/node_modules/@backstage/backend-defaults/src/entrypoints/httpAuth/httpAuthServiceFactory.ts:150:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
    },
    "request": {
        "method": "GET",
        "url": "/api/kubernetes/clusters"
    },
    "response": {
        "statusCode": 401
    }
}

Any ideas what I can still do in order to get rid of this error ? Many thanks in advance !

**Comment by famibo** [comment]
Interestingly enough this error occurs only with built-in Backstage backend plugins ! 

When moving our own internal plugin to the second backend, all works like a charm !!

Our plugin consists of two parts: frontend and backend.
When frontend plugin calls the backend (via fetchApi), it sets the authorization (bearer) token manually (which we retrieve from IdentityApi via getCredentials function). And as I wrote above it works also when the backend plugin runs in a separate backend, **but only after I followed your hint regarding discovery config for the auth backend plugin as mentioned above:**

discovery:
endpoints:
- target: http://localhost:7007/api/auth
plugins: [auth]

My assumption is that Backstage doesn't set  credentials / authentication token when calling (via fetchApi) backend plugins which results in this "AuthenticationError: Missing credentials" error. What do you think ? 

Is this what you mentioned writing "You have an option to override the fetchApi and provide your own implementation" ?

If yes, is there an example how to overrride the fetchApi ?

**Comment by famibo** [comment]
After having a deeper look into Backstage code, we were able to find a solution for the "AuthenticationError: Missing credentials" error. The solution is the **FetchMiddlewares.injectIdentityAuth** static method :-)

Basically it is about adding following to the apis.ts (in the frontend app):

  createApiFactory({
    api: fetchApiRef,
    deps: {
      configApi: configApiRef,
      identityApi: identityApiRef,
      discoveryApi: discoveryApiRef,
    },
    factory: ({ configApi, identityApi, discoveryApi }) => {
      return createFetchApi({
        middleware: [
          FetchMiddlewares.resolvePluginProtocol({
            discoveryApi,
          }),
          FetchMiddlewares.injectIdentityAuth({
            identityApi,
            config: configApi,
            urlPrefixAllowlist: [
              'http://localhost:7007',
              'http://localhost:7008'
            ],
          }),
        ],
      });
    },
  }),
  
  In the meanwhile I don't see this anymore like as bug, so you can close this one if you want.
  
  IMHO this is only about quality of the documentation which is rather poor in this point.

**Comment by benjdlambert** [comment]
Sorry was just about to reply with what you've posted last!

Yep that's the way you can solve it for now, it means that it's going to decorate the requests to those lists with the identity automatically.

It also means that you should be able to remove your code that manually adds the auth header when using the `fetchApi`.

LIke I mentioned earlier, I think that the longer term solution would be to use the `plugin://pluginName` protocol, but that usage isn't far spread yet as we've not really encouraged the usage.

Gonna close this one for now, but if you want to raise a PR with some recommended change to the docs we'd be very grateful :pray:

**Comment by famibo** [comment]
@benjdlambert: I've created a new PR for Backstage docs, adding some new content reflecting my lessons learned about separating backends: https://github.com/backstage/backstage/pull/27409
Maybe you would like to have a look into it as well. Thanks !
